lib/parsers/cr2_parser.rb

```diff
@@ -0,0 +1,173 @@
+    # and https://github.com/lclevy/libcraw2/blob/master/docs/cr2_poster.pdf
+    if0_offset = to_hex(tiff_header[4..7])
+
+    set_orientation(io, if0_offset)
+    set_resolution(io, if0_offset)
+    set_preview(io, if0_offset)
+    set_model_and_date(io, if0_offset)
+
+    # Check CanonAFInfo or CanonAFInfo2 tags in maker notes for width & height
+    exif_offset = parse_ifd(io, if0_offset, EXIF_OFFSET_TAG)
+
+    set_photo_info(io, exif_offset[0])
+
+    makernote_offset = parse_ifd(io, exif_offset[0], MAKERNOTE_OFFSET_TAG)
+    af_info = parse_ifd(io, makernote_offset[0], AF2INFO_TAG)
+
+    # Old Canon models have CanonAFInfo tags
+    # Newer models have CanonAFInfo2 tags instead
+    # See https://sno.phy.queensu.ca/~phil/exiftool/TagNames/Canon.html
+    if !af_info.nil?
+      parse_new_model(io, af_info[0], af_info[1])
+    else
+      af_info = parse_ifd(io, makernote_offset[0], AFINFO_TAG)
+      parse_old_model(io, af_info[0], af_info[1])
+    end
+
+    FormatParser::Image.new(
+      format: :cr2,
+      width_px: @width,
+      height_px: @height,
+      orientation: @orientation,
+      image_orientation: @image_orientation,
+      preview: parse_preview_image(io),
+      intrinsics: intrinsics
+    )
+  end
+
+  private
+
+  def parse_ifd(io, offset, searched_tag)
+    io.seek(offset)
+    entries_count = to_hex safe_read(io, 2)
+    entries_count.times do
+      entry = safe_read(io, 12)
+      tag_id = to_hex(entry[0..1])
+      length = to_hex(entry[4..7])
+      value = to_hex(entry[8..11])
+      return [value, length] if tag_id == searched_tag
+    end
+    nil
+  end
+
+  def to_hex(sequence)
+    sequence.reverse.unpack('H*').join.hex
+  end
+
+  def parse_new_model(io, offset, length)
+    io.seek(offset)
+    items = safe_read(io, length)
+    @width = to_hex(items[8..9])
+    @height = to_hex(items[10..11])
+  end
+
+  def parse_old_model(io, offset, length)
+    io.seek(offset)
+    items = safe_read(io, length)
+    @width = to_hex(items[4..5])
+    @height = to_hex(items[6..7])
+  end
+
+  def set_orientation(io, offset)
+    orient = parse_ifd(io, offset, PREVIEW_ORIENTATION_TAG).first
+    # Some old models do not have orientation info in TIFF headers
+    return if orient > 8
+    # EXIF orientation is an one based index
+    # http://sylvana.net/jpegcrop/exif_orientation.html
+    @orientation = FormatParser::EXIFParser::ORIENTATIONS[orient - 1]
+    @image_orientation = orient
+  end
+
+  def set_resolution(io, offset)
+    resolution_data = parse_unsigned_rational_data(io, offset, PREVIEW_RESOLUTION_TAG)
+    @resolution = resolution_data[0] / resolution_data[1]
+  end
+
+  def set_preview(io, offset)
+    @preview_offset = parse_ifd(io, offset, PREVIEW_IMAGE_OFFSET_TAG).first
+    @preview_byte_count = parse_ifd(io, offset, PREVIEW_IMAGE_BYTE_COUNT_TAG).first
+  end
+
+  def parse_preview_image(io)
+    return if @preview_byte_count > FormatParser::MAX_BYTES || @preview_offset > FormatParser::MAX_SEEKS
```

Comment:
See above re. preview - in this case this will also not quite work like you expect it. The MAX_BYTES constraint is for the overall amount the parser is permitted to read from the IO in total. If we have already read some bytes for other tags we are already under that limit actually, and even though here `@preview_offset` might be lower than `MAX_SEEKS` you will probably be below `MAX_SEEKS` already for this parser. So it will in fact cause an exception to be raised if we try to read MAX_BYTES at this stage - the limited IO simply won't let us.

The reason we've set it up this way (not have parsers bother with how much they are allowed to read) – is this: if we encourage it (checking the amount of data read in parsers, in every parser), every parser then needs to have it's own arithmetic for figuring out how much it did read already, or whether it is a child of another parser, there will be bugs in that arithmetic, and we might be increasing our possible attack surface since we handle user-supplied files here. Instead we set a hard limit for every parser in one place, once, and make it globally configurable.

Reply:
You are right, I haven't thought it that way. After moving preview data into intrinsics, ```parse_preview_image```method is not needed anymore.